subprojects/precondition-tester/build.gradle.kts

```
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+plugins {
+    id("gradlebuild.internal.java")
+}
+
+description = "Internal project testing and collecting information about all the test preconditions."
+
+// Special configuration containing all the precondition classes
+val precondition by configurations.creating {
+    description = "Configuration for all TestPrecondition classes dynamically loaded by the precondition tester"
+    isCanBeResolved = false
+    isCanBeConsumed = true
+}
+
+configurations {
+    // All test runtime configurations should extend from the "precondition" configuration
+    listOf(testRuntimeOnly, integTestRuntimeOnly, archTestRuntimeOnly, crossVersionTestRuntimeOnly).forEach {
```

```suggestion
    listOf(testRuntimeOnly, integTestRuntimeOnly, crossVersionTestRuntimeOnly).forEach {
```

🤔 Why wire the tests into `archTestRuntimeOnly` as well? Those do not support preconditions, so they shouldn't run the precondition tests, right?

---

I wasn't thinking about that `archTest` doesn't really need the preconditions...
In this case, I guess, https://github.com/gradle/gradle/blob/52c36b9ca82047c9731da374f00fd6027983b65f/subprojects/precondition-tester/build.gradle.kts#L67-L83 needs some modifications also, so we skip executing when `arch...` tests are being done.

---

Yeah, let's hook this into the test tasks we know about instead of every test task we find.

---

Currently worked around by not configuring `archTest`. 
Sadly it seems that the test doesn't have its own test class, so we have to resort to task name matching.

---

We could cover the regular `test` task and `DistributionTest`s. For the testing, we might want to check covered preconditions per task name, so we'd notice immediately when a new test task arises that we don't cover.